source/geod24/concurrency.d

```diff
@@ -1006,6 +1007,601 @@ protected:
+                ChannelQueueEntry q_ent = enqueueEntry(this.writeq, &val);
+
+                q_ent.sem.wait();
+                return q_ent.success;
+            }
+
+            return success;
+        }
+
+        /***********************************************************************
+
+            Try to write a message to Channel without blocking
+
+            tryWrite locks the Channel and writes a message if it is possible
+            to do so without blocking. If the write would block, it returns without
+            unlocking the Channel to allow caller to `enqueueEntry` without competing
+            for the lock again.
+
+            If the tryWrite is being executed in a select call, it tries to consume the
+            `caller_sel` with the given `caller_sel_id`. It only proceeds if it can
+            successfully consume the `caller_sel`
+
+            Param:
+                val = Message to write to channel
+                caller_sel = SelectState instace of the select call being executed
+                caller_sel_id = id of the select call being executed
+
+            Returns:
+                Success/Failure
+
+        ***********************************************************************/
+
+        private bool tryWrite () (auto ref T val, SelectState caller_sel = null, uint caller_sel_id = 0)
+        {
+            this.lock.lock();
+
+            if (this.closed)
+            {
+                this.lock.unlock();
+
+                if (caller_sel)
+                    caller_sel.tryConsume(caller_sel_id, false);
+
+                return false;
+            }
+
+            if (ChannelQueueEntry readq_ent = dequeueEntry(this.readq, caller_sel, caller_sel_id))
+            {
+                *(readq_ent.pVal) = val;
+                this.lock.unlock();
+                readq_ent.sem.notify();
+                return true;
+            }
+
+            if (this.max_size > 0 // this.max_size > 0 = buffered
+                        && this.buffer.length < this.max_size
+                        && (!caller_sel || caller_sel.tryConsume(caller_sel_id)))
+            {
+                this.buffer ~= val;
+                this.lock.unlock();
+                return true;
+            }
+
+            return false;
+        }
+
+        /***********************************************************************
+
+            Try to write a message to Channel without blocking and if it fails,
+            create a write queue entry using the given `sel_state` and `sel_id`
+
+            Param:
+                val = Message to write to channel
+                sel_state = SelectState instace of the select call being executed
+                sel_id = id of the select call being executed
+
+        ***********************************************************************/
+
+        void selectWrite() (T* val, SelectState sel_state, uint sel_id)
+        {
+            bool success = tryWrite(*val, sel_state, sel_id);
```

Comment:
```suggestion
            assert(val !is null);
            assert(sel_state !is null);
            bool success = tryWrite(*val, sel_state, sel_id);
```

Alternatively:

```
void selectWrite() (T* val, SelectState sel_state, uint sel_id)
in
{
    assert(val !is null);
    assert(sel_state !is null);
}
body
{
}
```

I think since recently you can also write `do` instead of `body`.